src/table/iterator.rs

```diff
@@ -189,6 +195,269 @@ impl BlockIterator {
+
+        if self.bpos > self.table.as_ref().offsets_length() {
+            self.err = Some(IteratorError::EOF);
+            return;
+        }
+
+        if self.block_iterator.as_ref().unwrap().data.is_empty() {
+            match self.table.as_ref().block(self.bpos, self.use_cache()) {
+                Ok(block) => {
+                    let block_iterator = self.get_block_iterator(block);
+                    block_iterator.seek_to_first();
+                    self.err = block_iterator.err.clone();
+                }
+                Err(err) => self.err = Some(IteratorError::Error(err.to_string())),
+            }
+        } else {
+            let bi = self.block_iterator.as_mut().unwrap();
+            bi.next();
+            if !bi.valid() {
+                self.bpos += 1;
+                bi.data.clear();
+                self.next_inner();
+            }
+        }
+    }
+
+    fn prev_inner(&mut self) {
+        self.err = None;
+
+        if self.block_iterator.as_ref().unwrap().data.is_empty() {
+            match self.table.as_ref().block(self.bpos, self.use_cache()) {
+                Ok(block) => {
+                    let block_iterator = self.get_block_iterator(block);
+                    block_iterator.seek_to_last();
+                    self.err = block_iterator.err.clone();
+                }
+                Err(err) => self.err = Some(IteratorError::Error(err.to_string())),
+            }
+        } else {
+            let bi = self.block_iterator.as_mut().unwrap();
+            bi.prev();
+            if !bi.valid() {
+                if self.bpos == 0 {
+                    self.bpos = std::usize::MAX;
+                    // At this point, the iterator must be reset before using.
+                    // If the user calls `next`, `bpos` will overflow and panic.
+                    self.err = Some(IteratorError::EOF);
+                    return;
+                }
+                self.bpos -= 1;
+                bi.data.clear();
+                self.prev_inner();
+            }
+        }
+    }
+
+    pub fn key(&self) -> &[u8] {
+        &self.block_iterator.as_ref().unwrap().key
+    }
+
+    pub fn value(&self) -> Value {
+        let mut value = Value::default();
+        value.decode(&self.block_iterator.as_ref().unwrap().val);
+        value
+    }
+
+    /// `next` points the iterator to next element.
+    /// Note that if the iterator becomes invalid after operation,
+    /// you must reset the iterator by using `rewind` or `seek`
+    /// before using it again.
+    pub fn next(&mut self) {
+        if self.opt & ITERATOR_REVERSED == 0 {
+            self.next_inner();
+        } else {
+            self.prev_inner();
+        }
+    }
+
+    /// `prev` points the iterator to previous element.
+    /// Note that if the iterator becomes invalid after operation,
+    /// you must reset the iterator by using `rewind` or `seek`
+    /// before using it again.
+    pub(crate) fn prev(&mut self) {
```

Comment:
seems upstream doesn't provide `prev` in the iterator interface.

Reply:
Table iterator provides a private `prev` for test use. FYI, https://github.com/dgraph-io/badger/blob/master/table/iterator.go#L331

Reply:
it's `prev_inner`